language/move-vm/runtime/src/session.rs

```diff
@@ -68,6 +70,27 @@ impl<'r, 'l, S: MoveResolver> Session<'r, 'l, S> {
         ty_args: Vec<TypeTag>,
         args: Vec<Vec<u8>>,
         gas_status: &mut GasStatus,
+    ) -> VMResult<Vec<Vec<u8>>> {
+        let mut no_ext = NativeContextExtensions::default();
+        self.execute_function_with_extensions(
+            module,
+            function_name,
+            ty_args,
+            args,
+            gas_status,
+            &mut no_ext,
+        )
+    }
+
+    /// Same as `execute_function`, but with native context extensions.
+    pub fn execute_function_with_extensions(
+        &mut self,
+        module: &ModuleId,
+        function_name: &IdentStr,
+        ty_args: Vec<TypeTag>,
+        args: Vec<Vec<u8>>,
+        gas_status: &mut GasStatus,
+        extensions: &mut NativeContextExtensions,
```

Comment:
Hmmm, I'm not sure we want to pass it in this way.

In the extensions here, I wanted to make it clear when to discard/keep the changes here. So I would have expected this being passed back alongside the execution result

Reply:
Maybe we could pass this in as an owned value? And then only pass it back alongside the change set on success?

Reply:
This makes only sense for the `execute_function_with_effects(self, ...)` version, and I did it for this one based on your feedback.

The other functions and this one are already designed for taking mutable references to handle state. Right now its not only `self`, but also `gas_status` which is treated like this. It's fitting the current APIs to use `&mut` extensions as well.

I hope we can move forward with this PR now. This is blocking other work, and I'm on time pressure as I'm going on a trip soon.

Reply:
> I hope we can move forward with this PR now. This is blocking other work, and I'm on time pressure as I'm going on a trip soon.

As you know today will be a bit of a chaotic day for me, but I'll try my best to get through this 

> The other functions and this one are already designed for taking mutable references to handle state. Right now its not only self, but also gas_status which is treated like this. It's fitting the current APIs to use &mut extensions as well.

Gas status is a bit different because it needs to be modifying the gas as the interpreter is running.
Similarly, the DataStore is also only mut for module publishing... and events... which it really shouldn't be (which I think long ago we were hoping to fix). There is no ability to otherwise write to global storage with that API, hence the &mut is a bit different there

This NativeContextExtensions is jsut way too general to pass in via &mut in my opinion. I understand you can still do non-transactional things if you start using refcells/rcs/etc, but I want to try to make it obvious what the correct behavior should be here. And I think that means passing it in as an owned value everywhere, and only returning it for functions that also produce a changeset

Reply:
I disagree and don't understand why you are holding up this PR and this nitty detail.

All those APIs work on mutable inputs. Why should extensions be different? What is special about them?

This is a minor stylistic issue you are bringing up here. If you aren't happy with the way how the current APIs are designed (i.e. working on `&mut) feel free to come back and refactor it.